src/mono/wasi/Wasi.Build.Tests/BuildTestBase.cs

```diff
@@ -0,0 +1,713 @@
+        }
+
+        public static string AddItemsPropertiesToProject(string projectFile, string? extraProperties=null, string? extraItems=null, string? atTheEnd=null)
+        {
+            if (extraProperties == null && extraItems == null && atTheEnd == null)
+                return projectFile;
+
+            XmlDocument doc = new();
+            doc.Load(projectFile);
+
+            XmlNode root = doc.DocumentElement ?? throw new Exception();
+            if (extraItems != null)
+            {
+                XmlNode node = doc.CreateNode(XmlNodeType.Element, "ItemGroup", null);
+                node.InnerXml = extraItems;
+                root.AppendChild(node);
+            }
+
+            if (extraProperties != null)
+            {
+                XmlNode node = doc.CreateNode(XmlNodeType.Element, "PropertyGroup", null);
+                node.InnerXml = extraProperties;
+                root.AppendChild(node);
+            }
+
+            if (atTheEnd != null)
+            {
+                XmlNode node = doc.CreateNode(XmlNodeType.DocumentFragment, "foo", null);
+                node.InnerXml = atTheEnd;
+                root.InsertAfter(node, root.LastChild);
+            }
+
+            doc.Save(projectFile);
+
+            return projectFile;
+        }
+
+        public void Dispose()
+        {
+            if (_projectDir != null && _enablePerTestCleanup)
+                _buildContext.RemoveFromCache(_projectDir, keepDir: s_skipProjectCleanup);
+        }
+
+        private static string GetEnvironmentVariableOrDefault(string envVarName, string defaultValue)
+        {
+            string? value = Environment.GetEnvironmentVariable(envVarName);
+            return string.IsNullOrEmpty(value) ? defaultValue : value;
+        }
+
+        internal BuildPaths GetBuildPaths(BuildArgs buildArgs, bool forPublish=true)
+        {
+            string objDir = GetObjDir(buildArgs.Config);
+            string bundleDir = Path.Combine(GetBinDir(baseDir: _projectDir, config: buildArgs.Config), "AppBundle");
+            string wasmDir = Path.Combine(objDir, "wasm", forPublish ? "for-publish" : "for-build");
+
+            return new BuildPaths(wasmDir, objDir, GetBinDir(buildArgs.Config), bundleDir);
+        }
+
+        internal IDictionary<string, FileStat> StatFiles(IEnumerable<string> fullpaths)
+        {
+            Dictionary<string, FileStat> table = new();
+            foreach (string file in fullpaths)
+            {
+                if (File.Exists(file))
+                    table.Add(Path.GetFileName(file), new FileStat(FullPath: file, Exists: true, LastWriteTimeUtc: File.GetLastWriteTimeUtc(file), Length: new FileInfo(file).Length));
+                else
+                    table.Add(Path.GetFileName(file), new FileStat(FullPath: file, Exists: false, LastWriteTimeUtc: DateTime.MinValue, Length: 0));
+            }
+
+            return table;
+        }
+
+        protected static string GetSkiaSharpReferenceItems()
```

Comment:
Does SkiaSharp run on wasi ?

Reply:
This, and other bits will get de-duplicated/cleaned up in a follow up PR.